playbooks/roles/web/templates/etc/varnish/default.vcl.j2

```diff
@@ -33,54 +33,110 @@ acl purge {
 # sub vcl_recv #
 ################
 sub vcl_hash {
-  hash_data(req.http.host);
-  hash_data(req.url);
+  # Hash protocol and URL.
   hash_data(req.http.X-Forwarded-Proto);
-  return(hash);
+  hash_data(req.url);
+
+  # Hash host or server IP.
+  if (req.http.host) {
+      hash_data(req.http.host);
+  } else {
+      hash_data(server.ip);
+  }
+
+  # Hash cookies for requests that have them.
+  if (req.http.Cookie) {
+      hash_data(req.http.Cookie);
+  }
+
+  return (hash);
 }
 
 ################
 # sub vcl_recv #
 ################
 sub vcl_recv {
+  # Only deal with "normal" method requests. (http://book.varnish-software.com/3.0/VCL_Basics.html#default-vcl-recv)
+  if (req.request != "GET" &&
+      req.request != "HEAD" &&
+      req.request != "PUT" &&
+      req.request != "POST" &&
+      req.request != "TRACE" &&
+      req.request != "OPTIONS" &&
+      req.request != "PATCH" &&
+      req.request != "DELETE") {
+      /* Non-RFC2616 or CONNECT which is weird. */
+      return (pipe);
+  }
+
+  # Only cache GET or HEAD requests. This makes sure the POST requests are always passed. (http://book.varnish-software.com/3.0/VCL_Basics.html#default-vcl-recv)
+  if (req.request != "GET" && req.request != "HEAD") {
+      return (pass);
+  }
+
+  # Normalize Accept-Encoding header.
+  if (req.http.Accept-Encoding) {
+      if (req.url ~ "^[^?]*\.(7z|avi|bz2|docx|eot|flac|flv|gif|ico|gz|jpeg|jpg|mka|mkv|mov|mp3|mp4|mpeg|mpg|ogg|ogm|opus|pdf|png|pptx|rar|svgz|tbz|tgz|txz|webm|webp|woff2|xlsx|xz|zip)(\?.*)?$") {
+          # No point in compressing these.
+          remove req.http.Accept-Encoding;
+      } elsif (req.http.Accept-Encoding ~ "gzip") {
+          set req.http.Accept-Encoding = "gzip";
+      } elsif (req.http.Accept-Encoding ~ "deflate") {
+          set req.http.Accept-Encoding = "deflate";
+      } else {
+          # Unkown algorithm.
+          remove req.http.Accept-Encoding;
+      }
+  }
+
+  # Large static files should be piped, so they are delivered directly to the end-user without
+  # waiting for Varnish to fully read the file first.
+  if (req.url ~ "^[^?]*\.(7z|avi|bz2|flac|flv|gz|mka|mkv|mov|mp3|mp4|mpeg|mpg|ogg|ogm|opus|rar|tar|tgz|tbz|txz|wav|webm|xz|zip)(\?.*)?$") {
+      unset req.http.Cookie;
+      return (pipe);
+  }
+
+  # Remove all cookies for static files.
+  if (req.url ~ "^[^?]*\.(bmp|css|csv|doc|docx|eot|gif|ico|jpeg|jpg|js|odt|otf|pdf|png|ppt|pptx|rtf|svg|svgz|swf|ttf|txt|webp|woff|woff2|xls|xlsx|xml)(\?.*)?$") {
+      unset req.http.Cookie;
+      return (lookup);
+  }
```

Comment:
We're already removing cookies from static files when inserting them to the cache, so we have to remove them when looking for them.